Source/Details/ASCollectionViewLayoutCalculating.h

```diff
@@ -0,0 +1,57 @@
+//
+//  ASCollectionViewLayoutCalculating.h
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 28/2/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <AsyncDisplayKit/ASDimension.h>
+
+@class ASLayout, ASElementMap;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol ASCollectionViewLayoutCalculating <NSObject>
+
+/**
+ * @abstract Return a layout of elements.
+ *
+ * @param size The size that the resulting layout should fit, usually equals to the size of the containing collection view.
+ *
+ * @param map An element map containings all elements to be laid out.
+ *
+ * @discussion This method can be called on multiple theads and thus must be thread-safe (and ideally, stateless).
+ * This method must block the calling thread but can dispatch to other theads to reduce the blocking time.
+ */
+- (ASLayout *)layoutThatFits:(CGSize)size forElementMap:(ASElementMap *)map;
+
+@end
+
+@protocol ASCollectionViewLayoutCalculatorProviding <NSObject>
+
+- (nullable id<ASCollectionViewLayoutCalculating>)collectionViewLayoutCalculator;
+
+- (CGSize)sizeForCalculatingCollectionViewLayout;
```

Comment:
Is there a way to use other terms in the API, like bounds or perhaps viewport?  Indeed this is not constrainedSize, but it kinda feels that way since the only places we ask for sizes from developers with layout are for constraints — so some way of disambiguating this would be nice.

boundsForCollectionLayout, viewportSizeForCollectionLayout, etc